### PR TITLE
[WIP] API support for pull requests

### PR DESCRIPTION
**Pull request by lstahlman**
This PR provides initial API support for pull requests (get, list, create, edit, merge).  See #2253 for the discussion regarding this portion of the API.

This is a work in progress and additional changes & documentation will be added.

This PR depends on gogits/go-gogs-client#53.  This also overlaps with PR #3547.

Here is the wiki documentation:
# Pull Requests
## List pull requests for a repository

```
GET /repos/:owner/:repo/pulls
```
##### Response

```
Status: 200 OK
Content-Type: application/json
```

``` json

[
{
  "id": 84,
  "number": 7,
  "user": {
    "id": 3,
    "username": "user3",
    "full_name": "Some User",
    "email": "user3@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "title": "Bugfix for XYZ",
  "body": "Fixes bug XYZ",
  "labels": [],
  "milestone": null,
  "assignee": null,
  "state": "closed",
  "comments": 0,
  "html_url": "http://example.com/user1/gogs/pulls/7",
  "mergeable": true,
  "merged": true,
  "merged_at": "2016-10-11T12:48:36-07:00",
  "merge_commit_sha": "a857e30a8af1bbbb43f065e1fc0fb1665036e95f",
  "merged_by": {
    "id": 1,
    "username": "user1",
    "full_name": "",
    "email": "user1@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "base": {
    "name": "devel",
    "repo_id": 25,
    "repo": {
      "id": 25,
      "owner": {
        "id": 1,
        "username": "user1",
        "full_name": "Some User",
        "email": "user1@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user1/gogs",
      "description": "",
      "private": false,
      "fork": false,
      "html_url": "http://example.com/user1/gogs",
      "ssh_url": "gogs@example.com:user1/gogs.git",
      "clone_url": "http://example.com/user1/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 1,
      "watchers_count": 1,
      "open_issues_count": 1,
      "default_branch": "master",
      "created_at": "2016-08-09T23:00:42-07:00",
      "updated_at": "2016-08-09T23:08:04-07:00"
    }
  },
  "head": {
    "name": "bugfix-XYZ",
    "repo_id": 131,
    "repo": {
      "id": 131,
      "owner": {
        "id": 3,
        "username": "user3",
        "full_name": "Some User",
        "email": "user3@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user3/gogs",
      "description": "",
      "private": false,
      "fork": true,
      "html_url": "http://example.com/user3/gogs",
      "ssh_url": "gogs@example.com/user3/gogs.git",
      "clone_url": "http://example.com/user3/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 0,
      "watchers_count": 1,
      "open_issues_count": 0,
      "default_branch": "master",
      "created_at": "2016-10-11T10:38:08-07:00",
      "updated_at": "2016-10-11T10:38:45-07:00"
    }
  },
  "merge_base": "66ef302b17e1d001043d8a9bb4fa93295a2d33c3"
}
]
```
## Get a single pull request

```
GET /repos/:owner/:repo/pulls/:id
```
##### Response

```
Status: 200 OK
Content-Type: application/json
```

``` json
{
  "id": 84,
  "number": 7,
  "user": {
    "id": 3,
    "username": "user3",
    "full_name": "Some User",
    "email": "user3@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "title": "Bugfix for XYZ",
  "body": "Fixes bug XYZ",
  "labels": [],
  "milestone": null,
  "assignee": null,
  "state": "closed",
  "comments": 0,
  "html_url": "http://example.com/user1/gogs/pulls/7",
  "mergeable": true,
  "merged": true,
  "merged_at": "2016-10-11T12:48:36-07:00",
  "merge_commit_sha": "a857e30a8af1bbbb43f065e1fc0fb1665036e95f",
  "merged_by": {
    "id": 1,
    "username": "user1",
    "full_name": "",
    "email": "user1@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "base": {
    "name": "devel",
    "repo_id": 25,
    "repo": {
      "id": 25,
      "owner": {
        "id": 1,
        "username": "user1",
        "full_name": "Some User",
        "email": "user1@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user1/gogs",
      "description": "",
      "private": false,
      "fork": false,
      "html_url": "http://example.com/user1/gogs",
      "ssh_url": "gogs@example.com:user1/gogs.git",
      "clone_url": "http://example.com/user1/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 1,
      "watchers_count": 1,
      "open_issues_count": 1,
      "default_branch": "master",
      "created_at": "2016-08-09T23:00:42-07:00",
      "updated_at": "2016-08-09T23:08:04-07:00"
    }
  },
  "head": {
    "name": "bugfix-XYZ",
    "repo_id": 131,
    "repo": {
      "id": 131,
      "owner": {
        "id": 3,
        "username": "user3",
        "full_name": "Some User",
        "email": "user3@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user3/gogs",
      "description": "",
      "private": false,
      "fork": true,
      "html_url": "http://example.com/user3/gogs",
      "ssh_url": "gogs@example.com/user3/gogs.git",
      "clone_url": "http://example.com/user3/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 0,
      "watchers_count": 1,
      "open_issues_count": 0,
      "default_branch": "master",
      "created_at": "2016-10-11T10:38:08-07:00",
      "updated_at": "2016-10-11T10:38:45-07:00"
    }
  },
  "merge_base": "66ef302b17e1d001043d8a9bb4fa93295a2d33c3"
}
```
## Create a pull request

```
POST /repos/:owner/:repo/pulls
```
##### Parameters

| Name | Type | Description |
| --- | --- | --- |
| head | string | **Required** Head branch |
| base | string | **Required** Base branch |
| title | string | **Required** Title of the pull request |
| body | string | **Required** Body of the pull request |
| assignee | string | Username for the user that this pull request should be assigned to. _NOTE: Only users with write access can set the assignee for new pull requests. The assignee is silently dropped otherwise._ |
| milestone | int | The **ID** of the milestone to associate this pull request with. _NOTE: Only users with write access can set the milestone for new pull requests. The milestone is silently dropped otherwise._ |
| labels | array of int | Label **IDs** to associate with this pull request. _NOTE: Only users with write access can set labels for new pull requests. Labels are silently dropped otherwise._ |
##### Example

``` json
{
  "head": "bugfix-XYZ",
  "base": "develop",
  "title": "Bugfix for XYZ",
  "body": "Fixes bug XYZ",
  "assignee": "user3",
  "milestone": 1,
  "labels": [2, 5]
}
```
##### Response

If a pull request already exists for the same head & base:

```
Status: 409 Conflict
```

Otherwise if the pull request is created:

```
Status: 201 Created
Content-Type: application/json
```

``` json
{
  "id": 84,
  "number": 7,
  "user": {
    "id": 3,
    "username": "user3",
    "full_name": "Some User",
    "email": "user3@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "title": "Bugfix for XYZ",
  "body": "Fixes bug XYZ",
  "labels": [],
  "milestone": null,
  "assignee": null,
  "state": "open",
  "comments": 0,
  "html_url": "http://example.com/user1/gogs/pulls/7",
  "mergeable": true,
  "merged": false,
  "merged_at": null,
  "merge_commit_sha": null,
  "merged_by": null,
  "base": {
    "name": "devel",
    "repo_id": 25,
    "repo": {
      "id": 25,
      "owner": {
        "id": 1,
        "username": "user1",
        "full_name": "Some User",
        "email": "user1@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user1/gogs",
      "description": "",
      "private": false,
      "fork": false,
      "html_url": "http://example.com/user1/gogs",
      "ssh_url": "gogs@example.com:user1/gogs.git",
      "clone_url": "http://example.com/user1/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 1,
      "watchers_count": 1,
      "open_issues_count": 1,
      "default_branch": "master",
      "created_at": "2016-08-09T23:00:42-07:00",
      "updated_at": "2016-08-09T23:08:04-07:00"
    }
  },
  "head": {
    "name": "bugfix-XYZ",
    "repo_id": 131,
    "repo": {
      "id": 131,
      "owner": {
        "id": 3,
        "username": "user3",
        "full_name": "Some User",
        "email": "user3@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user3/gogs",
      "description": "",
      "private": false,
      "fork": true,
      "html_url": "http://example.com/user3/gogs",
      "ssh_url": "gogs@example.com/user3/gogs.git",
      "clone_url": "http://example.com/user3/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 0,
      "watchers_count": 1,
      "open_issues_count": 0,
      "default_branch": "master",
      "created_at": "2016-10-11T10:38:08-07:00",
      "updated_at": "2016-10-11T10:38:45-07:00"
    }
  },
  "merge_base": "66ef302b17e1d001043d8a9bb4fa93295a2d33c3"
}
```
## Edit a pull request

Repository owners and users with write access can perform this action.

```
PATCH /repos/:owner/:repo/pulls/:index
```
##### Parameters

| Name | Type | Description |
| --- | --- | --- |
| title | string | **Required** Title of the pull request |
| body | string | **Required** Body of the pull request |
| assignee | string | Username for the user that this pull request should be assigned to. _NOTE: Only users with write access can set the assignee for new pull requests. The assignee is silently dropped otherwise._ |
| milestone | int | The **ID** of the milestone to associate this pull request with. _NOTE: Only users with write access can set the milestone for new pull requests. The milestone is silently dropped otherwise._ |
| labels | array of int | Label **IDs** to associate with this pull request. _NOTE: Only users with write access can set labels for new pull requests. Labels are silently dropped otherwise._ |

If a parameter is empty, that part of the pull request will remain unchanged
##### Example

``` json
{
  "body":"Fixes bug XYZ. Updated PR to reflect feedback."
}
```
##### Response

```
Status: 200 OK
Content-Type: application/json
```

``` json
{
  "id": 84,
  "number": 7,
  "user": {
    "id": 3,
    "username": "user3",
    "full_name": "Some User",
    "email": "user3@example.com",
    "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
  },
  "title": "Bugfix for XYZ",
  "body": "Fixes bug XYZ. Updated PR to reflect feedback.",
  "labels": [],
  "milestone": null,
  "assignee": null,
  "state": "open",
  "comments": 0,
  "html_url": "http://example.com/user1/gogs/pulls/7",
  "mergeable": true,
  "merged": false,
  "merged_at": null,
  "merge_commit_sha": null,
  "merged_by": null,
  "base": {
    "name": "devel",
    "repo_id": 25,
    "repo": {
      "id": 25,
      "owner": {
        "id": 1,
        "username": "user1",
        "full_name": "Some User",
        "email": "user1@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user1/gogs",
      "description": "",
      "private": false,
      "fork": false,
      "html_url": "http://example.com/user1/gogs",
      "ssh_url": "gogs@example.com:user1/gogs.git",
      "clone_url": "http://example.com/user1/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 1,
      "watchers_count": 1,
      "open_issues_count": 1,
      "default_branch": "master",
      "created_at": "2016-08-09T23:00:42-07:00",
      "updated_at": "2016-08-09T23:08:04-07:00"
    }
  },
  "head": {
    "name": "bugfix-XYZ",
    "repo_id": 131,
    "repo": {
      "id": 131,
      "owner": {
        "id": 3,
        "username": "user3",
        "full_name": "Some User",
        "email": "user3@example.com",
        "avatar_url": "https://secure.gravatar.com/avatar/30fa571d2f85bb1c5a5b574d5c5c5005"
      },
      "name": "gogs",
      "full_name": "user3/gogs",
      "description": "",
      "private": false,
      "fork": true,
      "html_url": "http://example.com/user3/gogs",
      "ssh_url": "gogs@example.com/user3/gogs.git",
      "clone_url": "http://example.com/user3/gogs.git",
      "website": "",
      "stars_count": 0,
      "forks_count": 0,
      "watchers_count": 1,
      "open_issues_count": 0,
      "default_branch": "master",
      "created_at": "2016-10-11T10:38:08-07:00",
      "updated_at": "2016-10-11T10:38:45-07:00"
    }
  },
  "merge_base": "66ef302b17e1d001043d8a9bb4fa93295a2d33c3"
}
```
## Merge a pull request

Only users with write access to a repository can perform this action.

```
POST /repos/:owner/:repo/pulls/:id/merge
```
##### Response

Success:

```
Status: 200 OK
```

Failed to auto-merge or PR already merged:

```
Status: 405 Method Not Allowed
```

Failure on attempt to merge a closed PR:

```
Status: 404 Not Found
```

All other errors:

```
Status: 500 Internal Server Error
```
## Check if a pull request is merged

```
GET /repos/:owner/:repo/pulls/:id/merge
```
##### Response

If merged:

```
Status: 204 No Content
```

If not merged:

```
Status: 404 Not Found
```

### CODE REVIEW

**Comment by stanhu** [comment]
Great! Would love to see some tests for this.

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+		Page:        ctx.QueryInt("page"),
+		State:       ctx.QueryTrim("state"),
+		SortType:    ctx.QueryTrim("sort"),
+		Labels:      ctx.QueryStrings("labels"),
+		MilestoneID: ctx.QueryInt64("milestone"),
+	})*/
+
+	prs, maxResults, err := models.PullRequests(ctx.Repo.Repository.ID, &models.PullRequestsOptions{
+		Page:  form.Page,
+		State: form.State,
```
Comment (bkcsoft) [inline routers/api/v1/repo/pull.go]: These are always query-parameters

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -291,6 +298,14 @@ func RegisterRoutes(m *macaron.Macaron) {
...
 				})
 				m.Get("/editorconfig/:filename", context.RepoRef(), repo.GetEditorconfig)
+				m.Group("/pulls", func() {
+					m.Combo("").Get(bind(api.ListPullRequestsOptions{}), repo.ListPullRequests).Post(reqRepoWriter(), bind(api.CreatePullRequestOption{}), repo.CreatePullRequest)
+					m.Group("/:index", func() {
+						m.Combo("").Get(repo.GetPullRequest).Patch(reqRepoWriter(), bind(api.EditPullRequestOption{}), repo.EditPullRequest)
+						m.Combo("/merge").Get(repo.IsPullRequestMerged).Post(reqRepoWriter(), repo.MergePullRequest)
+					})
+
+				}, mustAllowPulls, context.ReferencesGitRepo())
```
Comment (bkcsoft) [inline routers/api/v1/api.go]: shouldn't `mustAllowPulls` be a function-call?

Comment (lstahlman) [inline routers/api/v1/api.go]: I'm following the style for /issues, as it also places a similar condition mustAllowIssues. No point calling the endpoint functions if pull requests aren't enabled.

Comment (bkcsoft) [inline routers/api/v1/api.go]: That seems odd :laughing:

File: models/pull.go [hunk 3]
```diff
@@ -117,13 +117,21 @@ func (pr *PullRequest) LoadIssue() (err error) {
 func (pr *PullRequest) APIFormat() *api.PullRequest {
 
 	apiIssue := pr.Issue.APIFormat()
+	baseBranch, _ := pr.BaseRepo.GetBranch(pr.BaseBranch)
```
Comment (bkcsoft) [inline models/pull.go]: In case of an error, should we return `nil` instead?

**Comment by hhenkel** [comment]
@lstahlman are you still working on this?
Any plans on sending it to github.com/go-gitea/gitea master?

**Comment by strk** [comment]
Yup, would love to see this merged !
Please consider migrating it to https://github.com/go-gitea/gitea

**Comment by Unknwon** [comment]
Hi, is this PR completed?

**Comment by hultberg** [comment]
Is this pull request dead? This a feature I would love to see in gogs, as we can't fetch pull requests via the api at all now, only issues.

**Comment by strk** [comment]
@hultberg the feature is available in Gitea since version 1.0.0 (December 2016):
 https://github.com/go-gitea/gitea/issues/248

**Comment by hultberg** [comment]
@strk Why not provide this feature to the project gittea forked?

**Comment by strk** [comment]
@hultberg no reason why not. But we actually forked Gitea to make it easier for pull requests to be merged, so you'd better ask this question to @Unknwon who's the only one with permissions to accept contributions in Gogs.

**Comment by Fastidious** [comment]
@strk the problem is that you replied to his question on whether or not the pull was dead with a non-relevant answer. It doesn’t help @hultberg if the feature is available in Gitea. Evidently he is using `gogs`; are you coming here trying to recruit people? 😉

**Comment by Unknwon** [comment]
@strk 

1. Just like what you guys did about copy-paste from Gogs? That's easy to me, but I live in the hard way.
2. I'm OK with positive promotion, but if you guys keep spamming, I'll react.

**Comment by strk** [comment]
@Unknwon you're really welcome to copy-paste from Gitea.
We all just want better softare for everyone.

**Comment by Unknwon** [comment]
@strk ah, that's nice, thanks for your kind advice.

I'll start deleting everything I consider irrelevant to corresponding threads.

**Comment by Xjs** [comment]
What is the current state of this PR?

**Pull request closed without merging**